File: ImperatorToCK3Tests/ImperatorWorldTests/Genes/AccessoryGeneTemplateTests.cpp
#include "gtest/gtest.h"
#include <sstream>
#include "Imperator/Genes/AccessoryGeneTemplate.h"
#include "Imperator/Genes/WeightBlock.h"


TEST(ImperatorWorld_AccessoryGeneTemplateTests, ageSexWeightBlocksDefaultsToEmpty) {
	std::stringstream input;
	input << "=\n";
	input << "{\n";
	input << "}";

	const Imperator::AccessoryGeneTemplate geneTemplate(input);

	ASSERT_TRUE(geneTemplate.getAgeSexWeightBlocs().empty());
}

TEST(ImperatorWorld_AccessoryGeneTemplateTests, ageSexWeightBlocksCanBeLoaded) {
	std::stringstream input;
	input << "=\n";
	input << "{\n";
	input << "male={}\n";
	input << "female={}\n";
	input << "boy=male\n";
	input << "girl=female\n";
	input << "}";

	const Imperator::AccessoryGeneTemplate geneTemplate(input);

	ASSERT_EQ(4, geneTemplate.getAgeSexWeightBlocs().size());
}

TEST(ImperatorWorld_AccessoryGeneTemplateTests, ageSexWithBlocksAreProperlyCopied) {
	std::stringstream input;
	input << "=\n";
	input << "{\n";
	input << "male={ 6 = hoodie 8 = trousers }\n";
	input << "female={ 4 = skirt 6 = top }\n";
	input << "boy=male\n";
	input << "girl=female\n";
	input << "}";

	const Imperator::AccessoryGeneTemplate geneTemplate(input);

	ASSERT_EQ(6, geneTemplate.getAgeSexWeightBlocs().find("male")->second->getAbsoluteWeight("hoodie"));
	ASSERT_EQ(8, geneTemplate.getAgeSexWeightBlocs().find("male")->second->getAbsoluteWeight("trousers"));
	ASSERT_EQ(4, geneTemplate.getAgeSexWeightBlocs().find("female")->second->getAbsoluteWeight("skirt"));
	ASSERT_EQ(6, geneTemplate.getAgeSexWeightBlocs().find("female")->second->getAbsoluteWeight("top"));
	ASSERT_EQ(6, geneTemplate.getAgeSexWeightBlocs().find("boy")->second->getAbsoluteWeight("hoodie"));
	ASSERT_EQ(8, geneTemplate.getAgeSexWeightBlocs().find("boy")->second->getAbsoluteWeight("trousers"));
	ASSERT_EQ(4, geneTemplate.getAgeSexWeightBlocs().find("girl")->second->getAbsoluteWeight("skirt"));
	ASSERT_EQ(6, geneTemplate.getAgeSexWeightBlocs().find("girl")->second->getAbsoluteWeight("top"));
}

File: ImperatorToCK3.UnitTests/CommonUtils/Genes/AccessoryGeneTemplateTests.cs
using commonItems;
using ImperatorToCK3.CommonUtils.Genes;
using Xunit;

namespace ImperatorToCK3.UnitTests.CommonUtils.Genes {
	public class AccessoryGeneTemplateTests {
		[Fact]
		public void AgeSexWeightBlocksDefaultsToEmpty() {
			var reader = new BufferedReader(
				"= {}"
			);
			var geneTemplate = new AccessoryGeneTemplate(reader);

			Assert.Empty(geneTemplate.AgeSexWeightBlocks);
		}

		[Fact]
		public void AgeSexWeightBlocksCanBeLoaded() {
			var reader = new BufferedReader(
				"={\n" +
				"male={}\n" +
				"female={}\n" +
				"boy=male\n" +
				"girl=female\n" +
				"}"
			);
			var geneTemplate = new AccessoryGeneTemplate(reader);

			Assert.Equal(4, geneTemplate.AgeSexWeightBlocks.Count);
		}

		[Fact]
		public void AgeSexWithBlocksAreProperlyCopied() {
			var reader = new BufferedReader(
				"=\n" +
				"{\n" +
				"male={ 6 = hoodie 8 = trousers }\n" +
				"female={ 4 = skirt 6 = top }\n" +
				"boy=male\n" +
				"girl=female\n" +
				"}"
			);
			var geneTemplate = new AccessoryGeneTemplate(reader);

			Assert.Equal((uint)6, geneTemplate.AgeSexWeightBlocks["male"].GetAbsoluteWeight("hoodie"));
			Assert.Equal((uint)8, geneTemplate.AgeSexWeightBlocks["male"].GetAbsoluteWeight("trousers"));
			Assert.Equal((uint)4, geneTemplate.AgeSexWeightBlocks["female"].GetAbsoluteWeight("skirt"));
			Assert.Equal((uint)6, geneTemplate.AgeSexWeightBlocks["female"].GetAbsoluteWeight("top"));
			Assert.Equal((uint)6, geneTemplate.AgeSexWeightBlocks["boy"].GetAbsoluteWeight("hoodie"));
			Assert.Equal((uint)8, geneTemplate.AgeSexWeightBlocks["boy"].GetAbsoluteWeight("trousers"));
			Assert.Equal((uint)4, geneTemplate.AgeSexWeightBlocks["girl"].GetAbsoluteWeight("skirt"));
			Assert.Equal((uint)6, geneTemplate.AgeSexWeightBlocks["girl"].GetAbsoluteWeight("top"));
		}
	}
}
